include/async/then.hpp

```diff
@@ -22,47 +22,105 @@
+    };
+
+    template <typename... Fs>
+    using arities_t =
+        decltype(std::declval<stdx::tuple<Fs...>>()
+                     .fold_left(
+                         std::pair{stdx::tuple{},
+                                   std::integral_constant<std::size_t, 0>{}},
+                         []<typename Acc, typename F>(Acc, F) {
+                             using Offset = typename Acc::second_type;
+                             using A = arity_t<
+                                 F, Offset,
+                                 std::integral_constant<std::size_t, 0>>;
+                             return std::pair{
+                                 boost::mp11::mp_push_back<
+                                     typename Acc::first_type, A>{},
+                                 boost::mp11::mp_plus<Offset, A>{}};
+                         })
+                     .first);
+};
+
+template <typename... Fs> using arities_t = stdx::tuple<stdx::arity_t<Fs>...>;
+
+template <typename Arities, typename... Fs>
+using offsets_t = boost::mp11::mp_pop_back<boost::mp11::mp_partial_sum<
+    boost::mp11::mp_push_front<Arities, boost::mp11::mp_int<0>>,
+    boost::mp11::mp_int<0>, boost::mp11::mp_plus>>;
+
+struct void_t {};
+template <typename T>
+using nonvoid_result_t = std::bool_constant<not std::is_same_v<void_t, T>>;
+
+template <std::size_t Offset>
+constexpr auto invoke =
+    []<typename F, typename T, std::size_t... Is>(
+        F &&f, T &&t, std::index_sequence<Is...>) -> decltype(auto) {
+    if constexpr (std::is_void_v<std::invoke_result_t<
+                      F, decltype(std::forward<T>(
+                             t)[stdx::index<Offset + Is>])...>>) {
+        std::invoke(std::forward<F>(f),
+                    std::forward<T>(t)[stdx::index<Offset + Is>]...);
+        return void_t{};
+    } else {
+        return std::invoke(std::forward<F>(f),
+                           std::forward<T>(t)[stdx::index<Offset + Is>]...);
+    }
+};
+} // namespace detail
+
 template <typename R, typename... Fs> struct base_receiver : R {
     [[no_unique_address]] stdx::tuple<Fs...> fs;
 
-    struct void_t {};
-    template <typename T>
-    using nonvoid_result_t = std::bool_constant<not std::is_same_v<void_t, T>>;
-
     template <typename Tag, typename... Args> auto set(Args &&...args) -> void {
-        auto const exec = []<typename F, typename Arg>(F &&f, Arg &&arg) {
-            if constexpr (std::is_void_v<std::invoke_result_t<F, Arg>>) {
-                std::invoke(std::forward<F>(f), std::forward<Arg>(arg));
-                return void_t{};
-            } else {
-                return std::invoke(std::forward<F>(f), std::forward<Arg>(arg));
-            }
+        using arities =
+            typename detail::args<Args &&...>::template arities_t<Fs...>;
+        using offsets = typename detail::offsets_t<arities, Fs...>;
+
+        auto arg_tuple = stdx::tuple<Args &&...>{std::forward<Args>(args)...};
```

Comment:
**thought**: do we need a `stdx::forward_as_tuple`?

Reply:
It seems to be becoming useful, yes.